d2common/d2math/vector2.go

```diff
@@ -0,0 +1,367 @@
+}
+
+// Angle computes the angle in radians with respect
+// to the positive x-axis
+func (v *Vector2) Angle() *big.Float {
+	// HACK we should find a way to do this purely
+	// with big.Float
+	floatX, _ := v.X().Float64()
+	floatY, _ := v.Y().Float64()
+	floatAngle := math.Atan2(floatX, floatY)
+
+	if floatAngle < 0 {
+		floatAngle += 2.0 * math.Pi
+	}
+
+	return big.NewFloat(floatAngle)
+}
+
+// SetAngle sets the angle of this Vector2
+func (v *Vector2) SetAngle(angle *big.Float) d2interface.Vector2 {
+	return v.SetToPolar(angle, v.Length())
+}
+
+// Add to this Vector2 the components of the given Vector2
+func (v *Vector2) Add(src d2interface.Vector2) d2interface.Vector2 {
+	v.x.Add(v.x, src.X())
+	v.y.Add(v.y, src.Y())
+
+	return v
+}
+
+// Subtract from this Vector2 the components of the given Vector2
+func (v *Vector2) Subtract(src d2interface.Vector2) d2interface.Vector2 {
+	v.x.Sub(v.x, src.X())
+	v.y.Sub(v.y, src.Y())
+
+	return v
+}
+
+// Multiply this Vector2 with the components of the given Vector2
+func (v *Vector2) Multiply(src d2interface.Vector2) d2interface.Vector2 {
+	v.x.Mul(v.x, src.X())
+	v.y.Mul(v.y, src.Y())
+
+	return v
+}
+
+// Scale this Vector2 by the given value
+func (v *Vector2) Scale(s *big.Float) d2interface.Vector2 {
+	v.x.Sub(v.x, s)
+	v.y.Sub(v.y, s)
+
+	return v
+}
+
+// Divide this Vector2 by the given Vector2
+func (v *Vector2) Divide(src d2interface.Vector2) d2interface.Vector2 {
+	v.x.Quo(v.x, src.X())
+	v.y.Sub(v.y, src.Y())
+
+	return v
+}
+
+// Negate thex and y components of this Vector2
+func (v *Vector2) Negate() d2interface.Vector2 {
+	return v.Scale(big.NewFloat(negative1))
+}
+
+// Distance calculate the distance between this Vector2 and the given Vector2
+func (v *Vector2) Distance(src d2interface.Vector2) *big.Float {
+	dist := v.DistanceSq(src)
+
+	return dist.Sqrt(dist)
+}
+
+// DistanceSq calculate the distance suared between this Vector2 and the given
+// Vector2
+func (v *Vector2) DistanceSq(src d2interface.Vector2) *big.Float {
+	delta := src.Clone().Subtract(v)
+	deltaSq := delta.Multiply(delta)
+
+	return big.NewFloat(zero).Add(deltaSq.X(), deltaSq.Y())
+}
+
+// Length returns the length of this Vector2
+func (v *Vector2) Length() *big.Float {
+	xsq, ysq := v.LengthSq()
+
+	return xsq.Add(xsq, ysq)
+}
+
+func (v *Vector2) LengthSq() (*big.Float, *big.Float) {
+	clone := v.Clone()
+	x, y := clone.X(), clone.Y()
+
+	return x.Mul(x, x), y.Mul(y, y)
+}
+
+// SetLength sets the length of this Vector2
+func (v *Vector2) SetLength(length *big.Float) d2interface.Vector2 {
+	return v.Normalize().Scale(length)
+}
+
+// Normalize Makes the vector a unit length vector (magnitude of 1) in the same
+// direction.
+func (v *Vector2) Normalize() d2interface.Vector2 {
+	xsq, ysq := v.LengthSq()
+	length := big.NewFloat(zero).Add(xsq, ysq)
+	one := big.NewFloat(1.0)
+
+	if length.Cmp(one) > 0 {
+		length.Quo(one, length.Sqrt(length))
+
+		v.x.Mul(v.x, length)
+		v.y.Mul(v.y, length)
+	}
+
+	return v
```

Comment:
I think this might have been made clearer if we had a Length/Norm you don't need to sqrt. Also, I think Normalize should also normalize vector with a Norm of less than one. Making sure every input vector has a Magnitude or norm of 1 afterward.